Fix Slack App Install for other Workspaces

oliviasculley opened the pull request:
This attempts to fix the installation process when trying to add this bot to a new workspace, not sure if it'll work though...

oliviasculley committed:
attempt to fix install

allella commented:
Above my paygrade. Maybe @ThorntonMatthewD can give it a review.

ThorntonMatthewD commented on src/server.py:
```diff
@@ -106,6 +108,46 @@ async def rate_limit_check_api(
+@API.get("/slack/install")
+async def slack_install(req: Request):
+    """Generate an install button for new installation requests"""
+    del req
+    state = STATE_STORE.issue()
+    url = AUTHORIZE_URL_GENERATOR.generate(state)
+    return (
+        f'<a href="{html.escape(url)}">'
+        "<img "
+        'alt="Add to Slack" '
+        'height="40" '
+        'width="139" '
+        'src="https://platform.slack-edge.com/img/add_to_slack.png" '
+        'srcset="'
+        "https://platform.slack-edge.com/img/add_to_slack.png 1x, "
+        'https://platform.slack-edge.com/img/add_to_slack@2x.png 2x" '
+        "/></a>"
+    )
+
+
+@API.get("/slack/auth")
+async def slack_auth(req: Request):
+    """Used for new Slack app installs for other orgs"""
+    if "code" in req.args:
+        if STATE_STORE.consume(req.args["state"]):
+            client = WebClient()
+            client.oauth_v2_access(
+                client_id=os.environ.get("CLIENT_ID"),
+                client_secret=os.environ.get("CLIENT_SECRET"),
+                redirect_uri=None,
+                code=req.args["code"],
+            )
+            return "The HackGreenville API bot has been installed successfully!"
+    error = req.args["error"] if "error" in req.args else ""
+    raise HTTPException(
+        status_code=400,
+        detail=f"Something is wrong with the installation (error: {html.escape(error)})",
+    )
+
+
```
```python
@API.get("/slack/install")
async def slack_install():
    """Generate an install button for new installation requests"""
    state = STATE_STORE.issue()
    url = AUTHORIZE_URL_GENERATOR.generate(state)

    return Response(f"""
        <a href="{html.escape(url)}">
        <img alt="Add to Slack" height="40" width="139"
             src="https://platform.slack-edge.com/img/add_to_slack.png"
             srcset="https://platform.slack-edge.com/img/add_to_slack.png 1x,
                     https://platform.slack-edge.com/img/add_to_slack@2x.png 2x"
        />
        </a>
    """)


@API.get("/slack/auth")
async def slack_auth(code: str = "", state: str = "", error: str = ""):
    """Used for new Slack app installs for other orgs"""
    if code and STATE_STORE.consume(state):
        client = WebClient()
        client.oauth_v2_access(
            client_id=os.environ.get("CLIENT_ID"),
            client_secret=os.environ.get("CLIENT_SECRET"),
            redirect_uri=None,
            code=code,
        )
        return "The HackGreenville API bot has been installed successfully!"
    raise HTTPException(
        status_code=400,
        detail=f"Something is wrong with the installation (error: {html.escape(error)})",
    )
```

This is how I ended up tweaking it to make it work with FastAPI. I also had to import `Response` at the top (L19) for generating the HTML for the button.

ThorntonMatthewD committed:
Adjust the handling of query paramters in slack_auth to be compatible with FastAPI.

Also, adjusted the response construction for the Slack app installation button.

oliviasculley committed:
format with black

ThorntonMatthewD commented:
![Screenshot_2025-02-04_14-38-03](https://github.com/user-attachments/assets/739dab6f-de15-464f-9650-0586090df0c3)
Thank you for running a lint roll on the PR, @oliviasculley! For some reason whatever was being picked up in CI wasn't being caught locally..
